Userland/Utilities/uniq.cpp

```diff
@@ -5,75 +5,122 @@
+    else
+        file = TRY(Core::Stream::File::open(filename, Core::Stream::OpenMode::Read));
+    return TRY(Core::Stream::BufferedFile::create(file.release_nonnull()));
 }
 
 ErrorOr<int> serenity_main(Main::Arguments arguments)
 {
     TRY(Core::System::pledge("stdio rpath wpath cpath"));
 
-    char const* inpath = nullptr;
-    char const* outpath = nullptr;
+    StringView inpath;
+    StringView outpath;
+    bool duplicates_only = false;
+    bool unique_only = false;
+    bool ignore_case = false;
+    bool print_count = false;
+    unsigned int skip_chars = 0;
+    unsigned int skip_fields = 0;
+
     Core::ArgsParser args_parser;
+    args_parser.add_option(duplicates_only, "Only print duplicated lines", "repeated", 'd');
+    args_parser.add_option(unique_only, "Only print unique lines (default)", "unique", 'u');
+    args_parser.add_option(ignore_case, "Ignore case when comparing lines", "ignore-case", 'i');
+    args_parser.add_option(print_count, "Prefix each line by its number of occurences", "count", 'c');
+    args_parser.add_option(skip_chars, "Skip N chars", "skip-chars", 's', "N");
+    args_parser.add_option(skip_fields, "Skip N fields", "skip-fields", 'f', "N");
     args_parser.add_positional_argument(inpath, "Input file", "input", Core::ArgsParser::Required::No);
     args_parser.add_positional_argument(outpath, "Output file", "output", Core::ArgsParser::Required::No);
     args_parser.parse(arguments);
 
-    FILE* infile = get_stream(inpath, "r");
-    FILE* outfile = get_stream(outpath, "w");
-
-    struct linebuf buffers[2];
-    struct linebuf* previous = &(buffers[0]);
-    struct linebuf* current = &(buffers[1]);
-    bool first_run = true;
-    for (;;) {
-        errno = 0;
-        ssize_t rc = getline(&(current->buf), &(current->len), infile);
-        if (rc < 0 && errno != 0) {
-            perror("getline");
-            exit(1);
-        }
-        if (rc < 0)
-            break;
-        if (!first_run && strcmp(current->buf, previous->buf) == 0)
-            continue;
+    if (!unique_only && !duplicates_only)
+        unique_only = true;
+    else if (unique_only && duplicates_only) {
+        // Printing duplicated and unique lines shouldn't print anything
+        return 0;
+    }
+
+    if (outpath.is_empty() || outpath == "-"sv)
+        outpath = "/dev/stdout"sv;
+
+    auto infile = TRY(open_file_or_stdin(inpath));
+    auto outfile = TRY(Core::Stream::File::open(outpath, Core::Stream::OpenMode::Write));
+
+    size_t count = 0;
+    ByteBuffer previous_buf = TRY(ByteBuffer::create_uninitialized(1024));
+    ByteBuffer current_buf = TRY(ByteBuffer::create_uninitialized(1024));
 
-        fputs(current->buf, outfile);
+    StringView previous = TRY(infile->read_line(previous_buf));
+    StringView previous_to_compare = skip(previous, skip_chars, skip_fields);
+
+    while (TRY(infile->can_read_line())) {
+        StringView current = TRY(infile->read_line(current_buf));
```

Comment:
We should error in case we can't fit the full line. As far as I can tell, `read_line` just fills up the buffer completely if it can't read until the delimiter, but doesn't complain about that in any way.

Reply:
pretty sure the use of TRY() makes the error delegate
isn't there an easier way to have an automatically reallocating ByteBuffer?

Reply:
`TRY` delegates an error if one is returned, but `read_line` doesn't return an error if it can't fit the line into the buffer. It just copies as much as it can into the buffer and then continues as normal.

As for how to grow the buffer, if there is no `resize` method or something similar, you may just have to reassign the buffer that you are currently trying to write into.

Reply:
it seems like other programs using Stream::read_line don't resize the ByteBuffer themselves. let's wait for a better solution than having every program resize the bytebuffer themselves

Reply:
Either way, before this is merged there should either be a resolution to this, or a FIXME about truncating long lines.

Reply:
Actually, it doesn't truncate, it just returns EMSGSIZE (`Runtime error: message size (errno=55`)

Reply:
added FIXME

Reply:
but it looks like everything application using Core::Stream::File->read_line() has this problem

Reply:
This FIXME should be in the previous commit.